### PR TITLE
revert^2: experimental tls.external site setting

### PR DESCRIPTION
**Pull request by keegancsmith**
reverting the revert https://github.com/sourcegraph/sourcegraph/pull/7741

AWS Code Commit broke with two commits: https://github.com/sourcegraph/sourcegraph/pull/7651 and https://github.com/sourcegraph/sourcegraph/pull/7648

The first broke it because previously NewCertOpt would not be run for it, now it always would be. The second broke it further because that PR can't wrap non-http.Transport RoundTrippers. This PR introduces a marker interface to let us skip that.

The effect is AWS Code Commit won't respect the new tls.external site setting. This is fine for now, since the setting is experimental.

Test Plan: Run e2e tests on enterprise dev.

Part of https://github.com/sourcegraph/sourcegraph/issues/71

Fixes https://github.com/sourcegraph/sourcegraph/issues/7745
Fixes https://github.com/sourcegraph/sourcegraph/issues/7744
Fixes https://github.com/sourcegraph/sourcegraph/issues/7738

### CODE REVIEW

**Comment by tsenart** [comment]
> The effect is AWS Code Commit won't respect the new tls.external site setting. This is fine for now, since the setting is experimental.

What makes the aws code commit transport "unwrappable"?

**Comment by keegancsmith** [comment]
AWS code commit client has a custom RoundTripper which blacklists a certain domain to stop redirects. Its pretty hacky to be honest. It requires much more invansive changes to make it possible to layer its transport needs on top of other SG transport needs.

**Comment by tsenart** [comment]
What if we implemented a method like [`GetTransport`](https://godoc.org/github.com/aws/aws-sdk-go-v2/aws#BuildableHTTPClient.GetTransport) on `stubBadHTTPRedirectTransport`?

File: internal/httpcli/external.go [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+	"github.com/sourcegraph/sourcegraph/internal/conf"
+	"github.com/sourcegraph/sourcegraph/schema"
+)
+
+type externalTransport struct {
+	base *http.Transport
+
+	mu        sync.RWMutex
+	config    *schema.TlsExternal
+	effective *http.Transport
```
Comment (mrnugget) [inline internal/httpcli/external.go]: I have a hard time making sense of `effective`. Is that the cloned and _configured_ version of `base`?

File: internal/httpcli/external.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+func (t *externalTransport) RoundTrip(r *http.Request) (*http.Response, error) {
+	t.mu.RLock()
+	config, effective := t.config, t.effective
+	t.mu.RUnlock()
+
+	if current := conf.Get().ExperimentalFeatures.TlsExternal; current == nil {
+		return t.base.RoundTrip(r)
+	} else if reflect.DeepEqual(config, current) {
```
Comment (mrnugget) [inline internal/httpcli/external.go]:
```suggestion
	} else if !reflect.DeepEqual(config, current) {
		effective = t.update()
	}
```

and then you only need one `effective.RoundTrip(r)`?

I also think that it makes it clearer what happens here

File: internal/httpcli/external.go (hunk 1 above)
Comment (keegancsmith) [reply internal/httpcli/external.go]: yes. Suggestion for a better name for either?

File: internal/httpcli/external.go (hunk 2 above)
Comment (keegancsmith) [reply internal/httpcli/external.go]: I like it. Will update in a follow-up PR. Branch cut is gonna happen and CI is slow atm so won't make it if I do it now :)

**Comment by keegancsmith** [comment]
> What if we implemented a method like [`GetTransport`](https://godoc.org/github.com/aws/aws-sdk-go-v2/aws#BuildableHTTPClient.GetTransport) on `stubBadHTTPRedirectTransport`?

I implemented that, but it ended up being far too convoluted. There are issues around wanting to mutate the Transport, but still keep the RoundTrip leading to a much more complicated API. I opted for this simpler solution for now.

**Pull request merged**